bin/coqtop.ml

```diff
@@ -0,0 +1,124 @@
+open Stdune
+open Import
+
+let doc =
+  "Execute the Coq toplevel with the local configuration."
+
+let man =
+  [ `S "DESCRIPTION"
+  ; `P
+      {|$(b,dune coq top FILE -- ARGS) runs the Coq toplevel to process the
+          given $(b,FILE). The given arguments are completed according to the
+          local configuration. This is equivalent to running $(b,coqtop ARGS)
+          with a $(b,_CoqProject) file containing the local configurations
+          from the $(b,dune) files, but does not require maintaining a
+          $(b,_CoqProject) file.|}
+  ; `Blocks Common.help_secs
+  ]
+
+let info = Term.info "top" ~doc ~man
+
+let term =
+  let+ common = Common.term
+  and+ context =
+    Common.context_arg ~doc:{|Run the Coq toplevel in this build context.|}
+  and+ coqtop =
+    let doc = "Run the given toplevel command instead of the default." in
+    Arg.(value & opt string "coqtop" & info [ "toplevel" ] ~docv:"CMD" ~doc)
+  and+ coq_file_arg =
+    Arg.(required & pos 0 (some string) None (Arg.info [] ~docv:"COQFILE"))
+  and+ extra_args =
+    Arg.(value & pos_right 0 string [] (Arg.info [] ~docv:"ARGS"))
+  in
+  let config = Common.init common in
+  let coqtop, argv, env =
+    Scheduler.go ~common ~config (fun () ->
+        let open Fiber.O in
+        let* setup = Import.Main.setup () in
+        let* setup = Memo.Build.run setup in
+        let sctx = Import.Main.find_scontext_exn setup ~name:context in
+        let context = Dune_rules.Super_context.context sctx in
+        (* Checks on the argument file. *)
+        let source = Common.prefix_target common coq_file_arg in
```

Comment:
Indeed the whole logic in this file to determine the path of the .v file etc is a bit too obscure for my taste, maybe you could add a comment explaining what is going on here? See also Rudi's comment below.

Reply:
I fixed that concurrently: the most of it is trying to compute a relative path when the argument file is an absolute path (there is a comment for that now). I added that because editors (in particular emacs) will typically call `coqtop` with an absolute path. This is not absolutely necessary, since we could also do the logic of computing the root of the dune project from the editor, and do the path transformation there.